src/cdk/a11y/key-manager/tree-key-manager.ts

```diff
@@ -0,0 +1,559 @@
+        }
+
+        // NB: return here, in order to avoid preventing the default action of non-navigational
+        // keys or resetting the buffer of pressed letters.
+        return;
+    }
+
+    // Reset the typeahead since the user has used a navigational key.
+    this._pressedLetters = [];
+    event.preventDefault();
+  }
+
+  /** Index of the currently active item. */
+  getActiveItemIndex(): number | null {
+    return this._activeItemIndex;
+  }
+
+  /** The currently active item. */
+  getActiveItem(): T | null {
+    return this._activeItem;
+  }
+
+  /**
+   * Focus the initial element; this is intended to be called when the tree is focused for
+   * the first time.
+   */
+  onInitialFocus(): void {
+    this._focusFirstItem();
+  }
+
+  /** Focus the first available item. */
+  private _focusFirstItem(): void {
+    this.focusItem(this._findNextAvailableItemIndex(-1));
+  }
+
+  /** Focus the last available item. */
+  private _focusLastItem(): void {
+    this.focusItem(this._findPreviousAvailableItemIndex(this._items.length));
+  }
+
+  /** Focus the next available item. */
+  private _focusNextItem(): void {
+    this.focusItem(this._findNextAvailableItemIndex(this._activeItemIndex));
+  }
+
+  /** Focus the previous available item. */
+  private _focusPreviousItem(): void {
+    this.focusItem(this._findPreviousAvailableItemIndex(this._activeItemIndex));
+  }
+
+  /**
+   * Focus the provided item by index.
+   * @param index The index of the item to focus.
+   * @param options Additional focusing options.
+   */
+  focusItem(index: number, options?: {emitChangeEvent?: boolean}): void;
+  focusItem(item: T, options?: {emitChangeEvent?: boolean}): void;
+  focusItem(itemOrIndex: number | T, options?: {emitChangeEvent?: boolean}): void;
+  focusItem(itemOrIndex: number | T, options: {emitChangeEvent?: boolean} = {}) {
+    // Set default options
+    options.emitChangeEvent ??= true;
+
+    let index =
+      typeof itemOrIndex === 'number'
+        ? itemOrIndex
+        : this._items.findIndex(item => this._trackByFn(item) === this._trackByFn(itemOrIndex));
+    if (index < 0 || index >= this._items.length) {
+      return;
+    }
+    const activeItem = this._items[index];
+
+    // If we're just setting the same item, don't re-call activate or focus
+    if (
+      this._activeItem !== null &&
+      this._trackByFn(activeItem) === this._trackByFn(this._activeItem)
+    ) {
+      return;
+    }
+
+    this._activeItem = activeItem ?? null;
+    this._activeItemIndex = index;
+
+    if (options.emitChangeEvent) {
+      // Emit to `change` stream as required by TreeKeyManagerStrategy interface.
+      this.change.next(this._activeItem);
+    }
+    this._activeItem?.focus();
+    if (this._activationFollowsFocus) {
+      this._activateCurrentItem();
+    }
+  }
+
+  private _updateActiveItemIndex(newItems: T[]) {
+    const activeItem = this._activeItem;
+    if (activeItem) {
+      const newIndex = newItems.findIndex(
```

Comment:
nit: return immediately if there is no active item so you can avoid the extra indentation block